Mathlib/LinearAlgebra/Matrix/PosDef.lean

```diff
@@ -37,60 +42,243 @@ open scoped Matrix
+    {m : Type*} [Fintype m] (B : Matrix n m R) :
+    PosSemidef (Bᴴ * A * B) := by
+  constructor
+  · exact isHermitian_conjTranspose_mul_mul B hA.1
+  · intro x
+    simpa only [star_mulVec, dotProduct_mulVec, vecMul_vecMul] using hA.2 (mulVec B x)
+
+lemma mul_mul_conjTranspose_same {A : Matrix n n R} (hA : PosSemidef A)
+    {m : Type*} [Fintype m] (B : Matrix m n R):
+    PosSemidef (B * A * Bᴴ) := by
+  simpa only [conjTranspose_conjTranspose] using hA.conjTranspose_mul_mul_same Bᴴ
+
+theorem submatrix {M : Matrix n n R} (hM : M.PosSemidef) (e : m → n) :
     (M.submatrix e e).PosSemidef := by
-  refine' ⟨hM.1.submatrix e, fun x => _⟩
-  have : (M.submatrix (⇑e) e).mulVec x = (M.mulVec fun i : n => x (e.symm i)) ∘ e := by
-    ext i
-    dsimp only [(· ∘ ·), mulVec, dotProduct]
-    rw [Finset.sum_bij' (fun i _ => e i) _ _ fun i _ => e.symm i] <;>
-      simp only [eq_self_iff_true, imp_true_iff, Equiv.symm_apply_apply, Finset.mem_univ,
-        submatrix_apply, Equiv.apply_symm_apply]
-  rw [this]
-  convert hM.2 fun i => x (e.symm i) using 3
-  unfold dotProduct
-  rw [Finset.sum_bij' (fun i _ => e i) _ _ fun i _ => e.symm i] <;>
-  simp
+  classical
+  rw [(by simp : M = 1 * M * 1), submatrix_mul (he₂ := Function.bijective_id),
+    submatrix_mul (he₂ := Function.bijective_id), submatrix_id_id]
+  simpa only [conjTranspose_submatrix, conjTranspose_one] using
+    conjTranspose_mul_mul_same hM (Matrix.submatrix 1 id e)
 #align matrix.pos_semidef.submatrix Matrix.PosSemidef.submatrix
 
+protected lemma zero : PosSemidef (0 : Matrix n n R) :=
+  ⟨isHermitian_zero, by simp⟩
+
+protected lemma one [DecidableEq n] : PosSemidef (1 : Matrix n n R) :=
+  ⟨isHermitian_one, fun x => by
+    rw [one_mulVec]; exact Fintype.sum_nonneg fun i => star_mul_self_nonneg _⟩
+
+protected lemma pow [DecidableEq n] {M : Matrix n n R} (hM : M.PosSemidef) (k : ℕ) :
+    PosSemidef (M ^ k) :=
+  match k with
+  | 0 => .one
+  | 1 => by simpa using hM
+  | (k + 2) => by
+    rw [pow_succ', pow_succ]
+    simpa only [hM.isHermitian.eq] using (hM.pow k).mul_mul_conjTranspose_same M
+
+/-- The eigenvalues of a positive semi-definite matrix are non-negative -/
+lemma eigenvalues_nonneg [DecidableEq n] {A : Matrix n n 𝕜}
+    (hA : Matrix.PosSemidef A) (i : n) : 0 ≤ hA.1.eigenvalues i :=
+  (hA.re_dotProduct_nonneg _).trans_eq (hA.1.eigenvalues_eq _).symm
+
+theorem transpose {M : Matrix n n R} (hM : M.PosSemidef) : Mᵀ.PosSemidef := by
+  refine ⟨IsHermitian.transpose hM.1, fun x => ?_⟩
+  convert hM.2 (star x) using 1
+  rw [mulVec_transpose, Matrix.dotProduct_mulVec, star_star, dotProduct_comm]
+
+section sqrt
+
+variable [DecidableEq n] {A : Matrix n n 𝕜} (hA : PosSemidef A)
+
+/-- The positive semidefinite square root of a positive semidefinite matrix -/
+@[pp_dot]
+noncomputable def sqrt : Matrix n n 𝕜 :=
+  hA.1.eigenvectorMatrix * diagonal ((↑) ∘ Real.sqrt ∘ hA.1.eigenvalues) * hA.1.eigenvectorMatrixᴴ
```

Comment:
At risk of prolonging the review further; should we be skipping `sqrt` and going straight to defining `rpow`? What do you think @j-loreaux?

Reply:
Probably you already guessed how I would react, but for the record: I am against this in general, since it smacks of reinventing continuous functional calculus via the back door (where do we stop? `exp`? `tan`? `riemannZeta`?); and I am even more against trying to do it in this particular PR.

Reply:
I'm happy for the resolution in this PR to be "add a TODO comment" or "add a comment explaining why generalizing to rpow is a bad idea", rather than actually making the generalization; but I'm not qualified to decide between those options.

Reply:
(I'd argue that the square root plays a distinguished role because we care a lot about *bilinear forms* on vector spaces -- inner products and their generalisations -- and hence raising to the power 2, and also its inverse, has a distinguished role in the theory. Higher n-linear structures are much less frequently encountered. So that is why I argue that we _should_ have a special implementation of matrix square root, but _not_ of more general matrix powers.)

Reply:
I agree with David. There's no point reinventing the functional calculus (*especially* not as part of this PR). And indeed, square roots are somehow more important than other roots / functions. (This is a deep claim, but I believe it to be true.) For instance, using only that square roots exist, one can show that a closed two-sided ideal in a C⋆-algebra is selfadjoint.